aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -397,58 +420,34 @@ std::tuple<Tensor &,Tensor &> mode_out(const Tensor& self, int64_t dim, bool kee
-      // TODO: qscheme
-      return std::tuple<Tensor, Tensor>(at::_make_per_tensor_quantized_tensor(max,
-        self.q_scale(), self.q_zero_point()), max_indices);
+TORCH_IMPL_FUNC(max_out)
+(const Tensor& self,
+ int64_t dim,
+ bool keepdim,
+ const Tensor& values,
+ const Tensor& indices) {
+  NoNamesGuard guard;
+  if (self.numel() > 0) {
+    if (self.numel() == 1 && self.dim() == 0) {
+      values.fill_(self);
+      indices.fill_(0);
     } else {
-      Tensor max = at::empty({0}, self.options());
-      return at::native::max_out(self, dim, keepdim, max, max_indices);
+      auto& values_mut = const_cast<Tensor&>(values);
+      auto& indices_mut = const_cast<Tensor&>(indices);
+      max_stub(self.device().type(), values_mut, indices_mut, self, dim, keepdim);
     }
-}
-
-static std::tuple<Tensor &,Tensor &> max_out_impl(Tensor& max, Tensor& max_indices,
-                                                  const Tensor& self, int64_t dim, bool keepdim) {
-  TORCH_CHECK(self.device().is_cpu() || self.is_cuda(),
-              "max only supports CPU AND CUDA device type, got: ", self.device().type());
-  TORCH_CHECK(self.layout() == Layout::Strided,
-              "max only supports strided layout, got: ", self.layout());
-  TORCH_CHECK(self.device() == max.device(),
-              "expected device ", self.device(), " but got ",
-              max.device(), " for max values output");
-  TORCH_CHECK(self.device() == max_indices.device(),
-              "expected device ", self.device(), " but got ",
-              max_indices.device(), " for indices output");
-  dim = maybe_wrap_dim(dim, self.dim());
-  if (self.numel() == 0) {
-    zero_numel_tensor_resize(max, max_indices, self, dim, keepdim, "max()");
-    return std::tie(max, max_indices);
-  }
-  else if (_dimreduce_return_trivial_no_ident(max, self, dim, keepdim, "max")) {
-    // case where self.numel() == 1. The result does not need to be reshaped
-    // as a case of reduction in this case.
-    TORCH_CHECK(!self.is_complex(), "max does not support complex inputs.");
-    AT_ASSERT(max.dim() == 0);
-    max_indices.resize_({}).fill_(0);
-    return std::forward_as_tuple(max, max_indices);
-  } else {
-    max_stub(self.device().type(), max, max_indices, self, dim, keepdim);
-    return std::tuple<Tensor &,Tensor &>{max, max_indices};
   }
 }
 
-std::tuple<Tensor&,Tensor&> max_out(const Tensor& self, int64_t dim, bool keepdim, Tensor& max, Tensor& max_indices) {
-  auto result = [&]() {
-    NoNamesGuard guard;
-    return max_out_impl(max, max_indices, self, dim, keepdim);
-  }();
-  namedinference::propagate_names_for_reduction(max, self, dim, keepdim);
-  namedinference::propagate_names_for_reduction(max_indices, self, dim, keepdim);
-  return result;
+std::tuple<Tensor, Tensor> max(const Tensor& self, int64_t dim, bool keepdim) {
+  // CUDA and CPU dispatch keys are handled by the structured kernel implementation.
+  TORCH_INTERNAL_ASSERT(self.is_quantized());
```

Comment:
I think it's probably fine to take out the internal assert, since we don't assert dispatcher correctness in our other native kernels.

Also maybe name this kernel to something like `qmax`, since it's only used for quantized now?